gxa/src/main/java/uk/ac/ebi/atlas/trader/ExpressionAtlasExperimentTrader.java

```diff
@@ -52,33 +47,27 @@ public ExpressionAtlasExperimentTrader(ExperimentDAO experimentDAO,
             }
         }
         experimentCachesPerType = builder.build();
+        experimentTypes = CacheBuilder.newBuilder().build(new CacheLoader<Pair<String, String>, ExperimentType>() {
+            @Override
+            public ExperimentType load(Pair<String, String> p) throws Exception {
+                return experimentDAO.findExperiment(p.getLeft(), p.getRight()).getExperimentType();
+            }
+        });
     }
 
     @Override
-    public Experiment getPublicExperiment(String experimentAccession) {
-        ExperimentType experimentType;
+    public Experiment getExperiment(String experimentAccession, String accessKey) {
         try {
-            experimentType = publicExperimentTypesCache.getExperimentType(experimentAccession);
-            return getExperimentFromCache(experimentAccession, experimentType);
+            return getExperimentFromCache(experimentAccession, experimentTypes.get(Pair.of(experimentAccession, accessKey)));
         } catch (ExecutionException e) {
-            LOGGER.error(e.getMessage());
-            throw new ResourceNotFoundException("Failed to load <em>" + experimentAccession + "</em>.");
-        }
-        catch (UncheckedExecutionException | NullPointerException e) {
-            LOGGER.error(e.getMessage());
-            throw new ResourceNotFoundException("Experiment <em>" + experimentAccession + "</em> not found.");
+            throw new RuntimeException(e);
```

Comment:
`ResourceNotFoundException` returns a 404 in `...ExceptionHandlingController`, whereas this makes it return a generic 400.

Reply:
This is a really good point. I've changed the DAO to consistently throw the ResourceNotFoundException there, and the LoadingCache currently catches exceptions and rethrows them as UncheckedExecutionException - I'll try to do something about it